algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i

```diff
@@ -94,107 +72,203 @@ public:
-                                          int * indices, FPType * distances, const size_t nClasses, const size_t k, VoteWeights voteWeights,
-                                          int * trainLabel, NumericTable * testLabelTable)
+                                          const size_t trainBlockSize, const size_t startTestIdx, const size_t nTrain, DAAL_UINT64 resultsToEvaluate,
+                                          DAAL_UINT64 resultsToCompute, const size_t nClasses, const size_t k, VoteWeights voteWeights,
+                                          int * trainLabel, const NumericTable * trainTable, const NumericTable * testTable,
+                                          NumericTable * testLabelTable, NumericTable * indicesTable, NumericTable * distancesTable,
+                                          TlsMem<FPType, cpu> & tlsDistances, TlsMem<int, cpu> & tlsIdx, TlsMem<FPType, cpu> & tlsMaxs,
+                                          TlsMem<FPType, cpu> & tlsKDistances, TlsMem<int, cpu> & tlsKIndexes, TlsMem<int, cpu> & tlsVoting)
     {
-        services::Status s;
+        FPType * distancesBuff = tlsDistances.local();
+        DAAL_CHECK_MALLOC(distancesBuff);
 
-        daal::services::internal::TArray<FPType, cpu> tmpDistancesArr(blockSize * nTrain);
-        daal::services::internal::TArray<int, cpu> tmpIndicesArr(blockSize * nTrain);
-        FPType * tmpDistances = tmpDistancesArr.get();
-        int * tmpIndices      = tmpIndicesArr.get();
-        DAAL_CHECK_MALLOC(tmpDistances);
-        DAAL_CHECK_MALLOC(tmpIndices);
-        s |= distancesInstance->computeBatch(startTestIdx, blockSize, 0, nTrain, tmpDistances);
+        int * idx = tlsIdx.local();
+        DAAL_CHECK_MALLOC(idx);
 
-        for (size_t i = 0; i < blockSize; ++i)
+        FPType * maxs = tlsMaxs.local();
+        DAAL_CHECK_MALLOC(maxs);
+
+        FPType * kDistances = tlsKDistances.local();
+        DAAL_CHECK_MALLOC(kDistances);
+
+        int * kIndexes = tlsKIndexes.local();
+        DAAL_CHECK_MALLOC(kIndexes);
+
+        int * voting = tlsVoting.local();
+        DAAL_CHECK_MALLOC(voting);
+
+        service_memset_seq<FPType, cpu>(maxs, MaxVal<FPType>::get(), trainBlockSize);
+        service_memset_seq<FPType, cpu>(kDistances, MaxVal<FPType>::get(), blockSize * k);
+
+        const size_t i1    = startTestIdx;
+        const size_t i2    = startTestIdx + blockSize;
+        const size_t iSize = blockSize;
+
+        ReadRows<FPType, cpu> inDataRows(const_cast<NumericTable *>(testTable), i1, i2 - i1);
+        DAAL_CHECK_BLOCK_STATUS(inDataRows);
+        const FPType * const testData = inDataRows.get();
+
+        const size_t outBlockSize = trainBlockSize;
+        const size_t outRows      = nTrain;
+        const size_t nOutBlocks   = outRows / outBlockSize + (outRows % outBlockSize > 0);
+
+        for (size_t outBlock = 0; outBlock < nOutBlocks; outBlock++)
```

Comment:
Why not nest the parallel for?
If nTrain >> nTest  parallelism will be ineffective

Reply:
But in the general case, this should still give a win.